core/drivers/cbmem_console.c

```diff
@@ -0,0 +1,167 @@
+// SPDX-License-Identifier: BSD-2-Clause
+/*
+ * Copyright (c) 2023, Linaro Limited
+ */
+
+#include <compiler.h>
+#include <console.h>
+#include <drivers/cbmem_console.h>
+#include <io.h>
+#include <keep.h>
+#include <kernel/dt.h>
+#include <libfdt.h>
+#include <mm/core_mmu.h>
+#include <util.h>
+
+#define CURSOR_MASK (BIT(28) - 1)
+#define OVERFLOW BIT(31)
+
+/*
+ * Structures describing coreboot's in-memory descriptor tables. See
+ * <coreboot>/src/commonlib/include/commonlib/coreboot_tables.h for
+ * canonical implementation.
+ */
+
+struct cb_header {
+	char signature[4];
+	uint32_t header_bytes;
+	uint32_t header_checksum;
+	uint32_t table_bytes;
+	uint32_t table_checksum;
+	uint32_t table_entries;
+};
+
+#define CB_TAG_CBMEM_CONSOLE 0x17
+
+struct cb_entry {
+	uint32_t tag;
+	uint32_t size;
+	uint64_t uint64;
+};
+
+static struct cbmem_console_data cbmem_console;
+
+static void cbmem_console_flush(struct serial_chip *chip __unused)
+{
+}
+
+static int cbmem_console_getchar(struct serial_chip *chip __unused)
+{
+	return 0;
+}
+
+static bool cbmem_console_have_rx_data(struct serial_chip *chip __unused)
+{
+	return false;
+}
+
+static void cbmem_console_putc(struct serial_chip *chip, int ch)
+{
+	struct cbmem_console_data *pd =
+		container_of(chip, struct cbmem_console_data, chip);
+	if (pd->size == 0)
+		return;
+
+	if ((pd->console->cursor & CURSOR_MASK) + 1 >= pd->size) {
+		pd->console->cursor &= ~CURSOR_MASK;
+		pd->console->cursor |= OVERFLOW;
+		pd->console->body[0] = (uint8_t)(ch & 0xFF);
+	} else {
+		pd->console->body[pd->console->cursor & CURSOR_MASK] =
+			(uint8_t)(ch & 0xFF);
+		pd->console->cursor++;
+	}
+}
+
+static const struct serial_ops cbmem_console_ops = {
+	.flush = cbmem_console_flush,
+	.getchar = cbmem_console_getchar,
+	.have_rx_data = cbmem_console_have_rx_data,
+	.putc = cbmem_console_putc,
+};
+DECLARE_KEEP_PAGER(cbmem_console_ops);
+
+static paddr_t get_cbmem_console_from_coreboot_table(paddr_t table_addr,
+						     size_t table_size)
+{
+	struct cb_header *header;
+	void *ptr;
```

Comment:
`uint8_t *ptr` would avoid a couple of type casts.

Reply:
Done